### ISSUE TITLE
Should the skater be allowed to go underground?

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/energy-skate-park-basics#432]
Discovered as something that causes #430, but I find it a little strange that the skater is allowed to go below ground. We should handle the case in #430, but I am wondering if a broader issue is that the skater is allowed to go underground at all. @arouinfar @samreid, what do you think?
![capture](https://user-images.githubusercontent.com/6396244/43098578-b82dd5a4-8e7c-11e8-946b-b5785991d6a6.PNG)

**Comment by arouinfar** [comment]
@jessegreenberg I also find it odd that the skater can go underground. Is there a reasonable way to prevent this? If the red point under the skateboard gets to h < 0, does the skater just "fall" off and sit on the ground next to the track? This method wouldn't prevent all underground cases, such as when the skater is only partially underground.
![image](https://user-images.githubusercontent.com/8419308/43099577-d73a7882-8e7f-11e8-9b1b-9ffab8d27df0.png)


I think the bigger issue is what happens to the skater's energy when underground. In your screenshot in https://github.com/phetsims/energy-skate-park-basics/issues/432#issue-343765766, PE is greater than zero (since the red dot is above the ground). However, in https://github.com/phetsims/energy-skate-park-basics/issues/430#issue-343171015, the skater has negative PE, so the energy representations -- particularly the pie chart -- break down.

I'm more concerned with preventing negative PE than a partially-underground skater, as in the screenshot above.

**Comment by jessegreenberg** [comment]
> If the red point under the skateboard gets to h < 0, does the skater just "fall" off and sit on the ground next to the track?

I tried this, but found a few other cases where h can be less than zero while dragging a control point, mostly while dragging a control  point. The track bumps above ground at the end of the dragging interaction rather than during, which is why portions of the track can currently reach h < 0.

Can we ensure that the track rather than just the control points remain above zero while dragging?

**Comment by jessegreenberg** [comment]
That behaves strangely because it bumps up move
![ezgif com-resize](https://user-images.githubusercontent.com/6396244/43103292-d696a4ae-8e8a-11e8-9857-ca677f4530d1.gif)

**Comment by jessegreenberg** [comment]
This bug is only possible because the skater continues to move before the track is shifted above ground on release.

One option is that we could stop skater movement and reset energies while dragging the track. Sort of like how we "reset" the energy when the skater is grabbed. Is it important for skating to continue while  dragging the track?

Another option (not too sure yet how this would impact performance) would be to prevent the *track* rather than control points from going below ground, so dragging would be constrained to this:
![save-2](https://user-images.githubusercontent.com/6396244/43110299-66891d72-8ea8-11e8-9933-3a04f8f086b7.PNG)

Rather than this:
![save-1](https://user-images.githubusercontent.com/6396244/43110304-6c8a3e22-8ea8-11e8-86fc-2c945f417313.PNG)

Or maybe there is another option. @arouinfar what do you think?

**Comment by samreid** [comment]
There is code that bumps the track to always be above y=0.  We could change that to be y>=SKATER_HEIGHT, but then you could never make a parabola with y=0.

I don't think it's too confusing for the skater to pass in front of the ground--it is already somewhat unusual that the skater is hanging to the track like a roller-coaster, but hitting the ground would disrupt the motion and maybe distract from what the user is trying to see.

**Comment by jessegreenberg** [comment]
Is the issue great enough that it needs to be resolved prior to the release of the upcoming LoL/phet-io version?

**Comment by arouinfar** [comment]
> One option is that we could stop skater movement and reset energies while dragging the track. Sort of like how we "reset" the energy when the skater is grabbed. Is it important for skating to continue while dragging the track?

@jessegreenberg  resetting the skater's energy seems a bit inconvenient, particularly if a user is trying to fine-tune the track (for example, to made the loop a bit shorter).

> Another option (not too sure yet how this would impact performance) would be to prevent the track rather than control points from going below ground, so dragging a control point, track would be constrained to this:

That certainly sounds good, but I'd also be worried about performance. 

> Is the issue great enough that it needs to be resolved prior to the release of the upcoming LoL/phet-io version?

This seems like more of an edge case to me, so it wouldn't be critical for LoL, but not 100% sure about PhET-iO.

**Comment by ariel-phet** [comment]
@arouinfar @jessegreenberg @samreid here are my thoughts on this particular issue

1. I checked the java version of ESP, the skater can go underground in that one. So it would appear this behavior has been around for a LONG time, and I don't think it has caused any issues or concerns

2. To get the skater to have a negative PE take a very specific track arrangement, the skater skating on the "back side" of the track, the user moving the track while the skater is moving such that the skater can go under ground, and once the track is released, it pops back up to where the track is not below ground. (So I think a pretty specific edge case)

3. My experience with this sim is that tweaking the model/track algorithm can lead to lots of unintended consequences given the very sandbox nature of the sim. I think such tweaks should be carefully considered before investing the time.

Basically, I think this is a strong case of "it is not particularly broken, lets not fix it"

**Comment by arouinfar** [comment]
That seems reasonable to me @ariel-phet, given that creating a situation with negative PE is more of an edge case.

When height < 0, it looks like the the model assumes (sets?) PE = 0, and the total energy is unaffected. For ESPB, I think that's acceptable.
![espbunderground](https://user-images.githubusercontent.com/8419308/43233209-070d7902-9032-11e8-960e-813848019958.gif)

**Comment by jessegreenberg** [comment]
Thanks everyone, sounds good. Closing this issue.
